File: app/src/main/java/in/smslite/contacts/PhoneContactList.java
package in.smslite.contacts;

import android.net.Uri;
import android.os.Parcelable;
import android.text.TextUtils;
import android.util.Log;

import java.util.ArrayList;
import java.util.List;

import in.smslite.LogTag;


public class PhoneContactList extends ArrayList<PhoneContact> {
    private static final long serialVersionUID = 1L;

    public static PhoneContactList getByNumbers(Iterable<String> numbers, boolean canBlock) {
        PhoneContactList list = new PhoneContactList();
        for (String number : numbers) {
            if (!TextUtils.isEmpty(number)) {
                list.add(PhoneContact.get(number, canBlock));
            }
        }
        return list;
    }

    public static PhoneContactList getByNumbers(String semiSepNumbers,
                                           boolean canBlock,
                                           boolean replaceNumber) {
        PhoneContactList list = new PhoneContactList();
        for (String number : semiSepNumbers.split(";")) {
            if (!TextUtils.isEmpty(number)) {
                PhoneContact contact = PhoneContact.get(number, canBlock);
                if (replaceNumber) {
                    contact.setNumber(number);
                }
                list.add(contact);
            }
        }
        return list;
    }

    /**
     * Returns a ContactList for the corresponding recipient URIs passed in. This method will
     * always block to query provider. The given URIs could be the phone data URIs or tel URI
     * for the numbers don't belong to any contact.
     *
     * @param uris phone URI to create the ContactList
     */
    public static PhoneContactList blockingGetByUris(Parcelable[] uris) {
        PhoneContactList list = new PhoneContactList();
        if (uris != null && uris.length > 0) {
            for (Parcelable p : uris) {
                Uri uri = (Uri) p;
                if ("tel".equals(uri.getScheme())) {
                    PhoneContact contact = PhoneContact.get(uri.getSchemeSpecificPart(), true);
                    list.add(contact);
                }
            }
            final List<PhoneContact> contacts = PhoneContact.getByPhoneUris(uris);
            if (contacts != null) {
                list.addAll(contacts);
            }
        }
        return list;
    }

    /**
     * Returns a ContactList for the corresponding recipient ids passed in. This method will
     * create the contact if it doesn't exist, and would inject the recipient id into the contact.
     */
    public static PhoneContactList getByIds(String spaceSepIds, boolean canBlock) {
        PhoneContactList list = new PhoneContactList();
        for (RecipientIdCache.Entry entry : RecipientIdCache.getAddresses(spaceSepIds)) {
            if (entry != null && !TextUtils.isEmpty(entry.number)) {
                PhoneContact contact = PhoneContact.get(entry.number, canBlock);
                contact.setRecipientId(entry.id);
                list.add(contact);
            }
        }
        return list;
    }

    public int getPresenceResId() {
        // We only show presence for single contacts.
        if (size() != 1)
            return 0;

        return get(0).getPresenceResId();
    }

    public String formatNames(String separator) {
        String[] names = new String[size()];
        int i = 0;
        for (PhoneContact c : this) {
            names[i++] = c.getName();
        }
        return TextUtils.join(separator, names);
    }

    public String formatNamesAndNumbers(String separator) {
        String[] nans = new String[size()];
        int i = 0;
        for (PhoneContact c : this) {
            nans[i++] = c.getNameAndNumber();
        }
        return TextUtils.join(separator, nans);
    }

    public String serialize() {
        return TextUtils.join(";", getNumbers());
    }

    public boolean containsEmail() {
        for (PhoneContact c : this) {
            if (c.isEmail()) {
                return true;
            }
        }
        return false;
    }

    public String[] getNumbers() {

        List<String> numbers = new ArrayList<>();
        String number;
        for (PhoneContact c : this) {
            number = c.getNumber();

            // Don't add duplicate numbers. This can happen if a contact name has a comma.
            // Since we use a comma as a delimiter between contacts, the code will consider
            // the same recipient has been added twice. The recipients UI still works correctly.
            // It's easiest to just make sure we only send to the same recipient once.
            if (!TextUtils.isEmpty(number) && !numbers.contains(number)) {
                numbers.add(number);
            }
        }
        return numbers.toArray(new String[numbers.size()]);
    }

    @Override
    public boolean equals(Object obj) {
        try {
            PhoneContactList other = (PhoneContactList)obj;
            // If they're different sizes, the contact
            // set is obviously different.
            if (size() != other.size()) {
                return false;
            }

            // Make sure all the individual contacts are the same.
            for (PhoneContact c : this) {
                if (!other.contains(c)) {
                    return false;
                }
            }

            return true;
        } catch (ClassCastException e) {
            return false;
        }
    }

    private void log(String msg) {
        Log.d(LogTag.TAG, "[ContactList] " + msg);
    }
}

File: app/src/main/java/in/smslite/contacts/PhoneContactList.kt
package `in`.smslite.contacts

import `in`.smslite.LogTag
import `in`.smslite.contacts.PhoneContact.Companion.getByPhoneUris
import android.net.Uri
import android.os.Parcelable
import android.text.TextUtils
import android.util.Log
import java.util.*

class PhoneContactList : ArrayList<PhoneContact?>() {
    // We only show presence for single contacts.
    val presenceResId: Int
        get() =// We only show presence for single contacts.
            if (size != 1) 0 else get(0)!!.presenceResId

    fun formatNames(separator: String?): String {
        val names = arrayOfNulls<String>(size)
        var i = 0
        for (c in this) {
            names[i++] = c?.name
        }
        return TextUtils.join(separator!!, names)
    }

    fun formatNamesAndNumbers(separator: String?): String {
        val nans = arrayOfNulls<String>(size)
        var i = 0
        for (c in this) {
            nans[i++] = c?.nameAndNumber
        }
        return TextUtils.join(separator!!, nans)
    }

    fun serialize(): String {
        return TextUtils.join(";", numbers)
    }

    fun containsEmail(): Boolean {
        for (c in this) {
            if (c?.isEmail == true) {
                return true
            }
        }
        return false
    }

    // Don't add duplicate numbers. This can happen if a contact name has a comma.
    // Since we use a comma as a delimiter between contacts, the code will consider
    // the same recipient has been added twice. The recipients UI still works correctly.
    // It's easiest to just make sure we only send to the same recipient once.
    val numbers: Array<String?>
        get() {
            val numbers: MutableList<String?> = ArrayList()
            var number: String
            for (c in this) {
                number = c?.number.toString()

                // Don't add duplicate numbers. This can happen if a contact name has a comma.
                // Since we use a comma as a delimiter between contacts, the code will consider
                // the same recipient has been added twice. The recipients UI still works correctly.
                // It's easiest to just make sure we only send to the same recipient once.
                if (!TextUtils.isEmpty(number) && !numbers.contains(number)) {
                    numbers.add(number)
                }
            }
            return numbers.toTypedArray()
        }

    override fun equals(obj: Any?): Boolean {
        return try {
            val other = obj as PhoneContactList?
            // If they're different sizes, the contact
            // set is obviously different.
            if (size != other!!.size) {
                return false
            }

            // Make sure all the individual contacts are the same.
            for (c in this) {
                if (!other.contains(c)) {
                    return false
                }
            }
            true
        } catch (e: ClassCastException) {
            false
        }
    }

    private fun log(msg: String) {
        Log.d(LogTag.TAG, "[ContactList] $msg")
    }

    companion object {
        private const val serialVersionUID = 1L
        fun getByNumbers(numbers: Iterable<String?>, canBlock: Boolean): PhoneContactList {
            val list = PhoneContactList()
            for (number in numbers) {
                if (!TextUtils.isEmpty(number)) {
                    list.add(PhoneContact[number!!, canBlock])
                }
            }
            return list
        }

        fun getByNumbers(semiSepNumbers: String,
                         canBlock: Boolean,
                         replaceNumber: Boolean): PhoneContactList {
            val list = PhoneContactList()
            for (number in semiSepNumbers.split(";").toTypedArray()) {
                if (!TextUtils.isEmpty(number)) {
                    val contact = PhoneContact[number, canBlock]
                    if (replaceNumber) {
                        contact.number(number)
                    }
                    list.add(contact)
                }
            }
            return list
        }

        /**
         * Returns a ContactList for the corresponding recipient URIs passed in. This method will
         * always block to query provider. The given URIs could be the phone data URIs or tel URI
         * for the numbers don't belong to any contact.
         *
         * @param uris phone URI to create the ContactList
         */
        fun blockingGetByUris(uris: Array<Parcelable>?): PhoneContactList {
            val list = PhoneContactList()
            if (uris != null && uris.size > 0) {
                for (p in uris) {
                    val uri = p as Uri
                    if ("tel" == uri.scheme) {
                        val contact = PhoneContact[uri.schemeSpecificPart, true]
                        list.add(contact)
                    }
                }
                val contacts = getByPhoneUris(uris)
                if (contacts != null) {
                    list.addAll(contacts)
                }
            }
            return list
        }

        /**
         * Returns a ContactList for the corresponding recipient ids passed in. This method will
         * create the contact if it doesn't exist, and would inject the recipient id into the contact.
         */
        fun getByIds(spaceSepIds: String?, canBlock: Boolean): PhoneContactList {
            val list = PhoneContactList()
            for (entry in RecipientIdCache.getAddresses(spaceSepIds)) {
                if (entry != null && !TextUtils.isEmpty(entry.number)) {
                    val contact = PhoneContact[entry.number, canBlock]
                    contact.recipientId = entry.id
                    list.add(contact)
                }
            }
            return list
        }
    }
}